### DISCUSSION TITLE
⚠️ Breaking changes tracker

### DISCUSSION THREAD

**Discussion by pi0** [nuxt/nuxt#16822] [General]
Until Nuxt 3 project reaches to release candidate stage with proper release notes, there might be unannounced breaking changes this helps us to iterate faster by trying changes and shape the best final API for Nuxt 3.

Since `nuxt3` npm package is auto-released and there are no release notes, this discussion is to keep track of the changelog of noticeable or breaking changes. If you would like to stay updated, please consider watching the repository or subscribing to this discussion.

#### #4449 `nuxt3` renamed to `nuxt` internally

In order to keep using edge releases, consider these changes:

1. Update `package.json`

```diff
{
  "devDependencies": {
-    "nuxt3": "latest"
+    "nuxt": "npm:nuxt3@latest"
  }
}
```

2. Update `nuxt.config`:

```diff
- import { defineNuxtConfig } from 'nuxt3'
+ import { defineNuxtConfig } from 'nuxt'

export default defineNuxtConfig({ ... })
```

----

#### #4422 nested routes suspense behaviour

The `page:start` and `page:finish` hooks are now only called _once_, with the top-level route component rather than any nested route components.

----

#### #4254, #4267 `runtimeConfig` namespace 

`publicRuntimeConfig` and `privateRuntimeConfig` are merged and simplified into new `runtimeConfig` option.

Old configuration:

```js
export default defineNuxtConfig({
  privateRuntimeConfig: {
    secretKey: ''
  },
  publicRuntimeConfig: {
    myValue: 123
  }
})
```

New configuration

```js
export default defineNuxtConfig({
  runtimeConfig: {
    secretKey: '',
    public: {
      myValue: true
    }
  }
})
```

**Usage:**

```js
export default defineNuxtPlugin(() => {
  console.log(useRuntimeConfig().secretKey) // only available on server-side and API routes
  console.log(useRuntimeConfig().public.myValue)
})
```

**Overriding at runtime**:

```bash
NUXT_PUBLIC_MY_VALUE=456 node .output/server/index.mjs
NUXT_SECRET_KEY=foobar node .output/server/index.mjs
```

----

#### #4218 [bridge] `@vueuse/head` is an optional dependency

**Note:** This is only applicable to nuxt bridge usage and not nuxt 3!

If you are using the experimental Bridge meta integration, you will need to explicitly add `@vueuse/head` to your dependencies. If you are using yarn or pnpm, you should also use resolutions to set your vue version to 2:

```json
{
  "resolutions": {
    "vue": "^2.6.14"
  }
}
```

----

#### #4019 Default cookie path

Default cookie path is now `/`. This is not a breaking change but might have security implications if you was depending on default being same as path.

----

#### #3956 **Migration to nitropack**

There should be normally no direct breaking changes but h3 and server APIs are updated. Please refer to the [unjs/nitropack documentation](https://github.com/unjs/nitro) for more info and open an issue if having a regression by this change.

***Major breaking changes:**

- Server API routes have a new format with `defineEventHandler`. Please consider migrating as old syntax will be unsupported in the future releases `https://v3.nuxtjs.org/guide/features/server-routes`
- [bridge] `nuxi generate` Uses new generate feature from nitro. You can use `nuxt generate` in order to use Nuxt 2 generator, however, this is not recommended!

----

#### #4066 Rename `useMeta` to `useHead`

Rename `useMeta` to `useHead` anywhere it occurs in your code, and use `app.head` in your `nuxt.config` rather than `meta`. Any imports of `useHead` should be from `#imports` rather than `#meta`.

----

#### #3985 Use payload cache for initial data fetching by default

- No longer refresh `useAsyncData` and `useFetch` calls if these have already been made once.

----

#### #3717 Use individual config layers for extending

- Internal option `_extends` renamed to `_layouts`.
- Any modifications to `nuxt.options.components.dirs` and `nuxt.options.autoImports.dirs` will be discarded. Instead use `addComponentsDir`/`addComponentsDirs` or directly `components:dirs`/`autoImports:dirs` hooks to extend them.

----

#### #3610: enable using `<NuxtLayout>` without pages integration

If you are using the `pages:layouts:extend` hook, then you can switch to using `app:resolve`, where you will be able to access `app.layouts.`

----

#### #2812: Split out webpack and postcss build options

If you are using webpack, you will need to update your `nuxt.config`:

```diff
  export default defineNuxtConfig({
-   vite: false,
+   builder: 'webpack'
  })
```

You will also now need to install the webpack builder separately. Add this to `package.json`:

```json
{
  "devDependencies": {
    "@nuxt/webpack-builder": "npm:@nuxt/webpack-builder-edge@latest"
  }
}
```

If you have configured your postcss plugins separately, you will now need to do so through a new top-level config option. This will apply to both webpack & vite builders.

```diff
  import { defineNuxtConfig } from 'nuxt3'

  export default defineNuxtConfig({
-   build: {
-     postcss: {
-       postcssOptions: {
-         plugins: {
-           'postcss-import': {}
-         }
-       }
-     }
-   },
+   postcss: {
+     plugins: {
+       'postcss-import': {}
+     }
+   }
  })
```

----

#### #3305: Disable global components by default

If you were relying on global registration of your components - for example, because you wanted to use a dynamically computed component name (e.g. `<component :is="'icon-' + myIcon" />`) then you can re-enable the previous behaviour by setting `components.global`:

```diff
  import { defineNuxtConfig } from 'nuxt3'
  
  export default defineNuxtConfig({
    components: {
+     global: true,
+     dirs: ['~/components']
    },
  })
```

---

#### #3268: Custom response type for `useFetch` using first generic

If you was already using `useFetch` or `useLazyFetch` with custom type, it needs to be updated:

```diff
--- useFetch<string, ResponseType>(....)
+++ useFetch<ResponseType>(....)
```

---

#### #3108: Multi layer `components/` dir support

`level` option in `components[]` is no longer supported. 

-----

#### #2976: key routes by interpolated path

- If you are using `<NuxtNestedPage>` or `<NuxtChildPage>` you should now replace these with `<NuxtPage` directly.
- Update any `child-key` props to `page-key` instead.
- If you are relying on the previous default unkeyed behavior of dynamic routes,  you can manually re-enable this:
   ```js
   definePageMeta({ key: false })
   ```
-----

#### #3011:  extract `<NuxtLayout>` from `<NuxtPage>`

* Add `<NuxtLayout>` to `app.vue` if you are using layouts:
   ```diff
    <template>
   +  <NuxtLayout>
        <NuxtPage />
   +  </NuxtLayout>
    </template>
   ```

* If you were passing layout manually to `<NuxtPage>`, move this to `<NuxtLayout>` instead:
   ```diff
    <template>
   +  <NuxtLayout name="my-layout">
   -     <NuxtPage layout="my-layout" />
   +     <NuxtPage />
   +  </NuxtLayout>
    </template>
   ```

* If you have been setting `layout: false` and manually using `<NuxtLayout>` without an explicit layout, you now need to provide a name option:

   ```diff
    <template>
   -  <NuxtLayout>
   +  <NuxtLayout name="default">
        <!-- your content -->
      </NuxtLayout>
    </template>
    
    <script setup>
    definePageMeta({
      layout: false
    })
    </script>
   ```

---

#### #2861: `transition` option of `definePageMeta` renamed to `pageTransition`

Please see #2861 for more information.

---

#### #2774: Only top-level files and subdirectory indices in `plugins/` are auto-registered

If you were previously depending on this feature, please use the `plugins` option in `nuxt.config` to manually register plugins or move them to the top level. A workaround is mentioned in https://github.com/nuxt/framework/issues/2780#issuecomment-1016969323 if you prefer to keep old behavior of scanning nested plugin dirs.

---

#### #2655: Options for `publicPath`, `router.base`, and `app` updated

Please see #2655 for more information.

---

#### #2275: `defineNuxtModule` API updated

Please see #2275 for more information.

**Comment by davidfrgla** [comment]
I followed your  migration guide and now I'm getting `ReferenceError: process is not defined`
(using [@coinbase/wallet-sdk](https://github.com/coinbase/coinbase-wallet-sdk))

**Reply by danielroe** [reply]
Would you open an issue instead of commenting here please? 🙏

**Reply by davidfrgla** [reply]
Sorry for that, I have already opened an issue #3858 that may be related

**Reply by SVV-team** [reply]
@thehedgeguy 
https://github.com/browserify/node-util/issues/43#issuecomment-1046110526

Its resolve ReferenceError: process is not defined

**Reply by davidfrgla** [reply]
Thanks, that solved it

**Comment by henningko** [comment]
Would it be possible to add a timestamp (or version?) to the description of each breaking change? I found that upgrading my package broke my app—after a longer period of not upgrading—and date/version would allow to quickly narrow on what might be causing the break. (I understand it's possible to reconstruct via the change history of the post, but would be way more handy if it's mentioned alongside the description).

**Reply by eMeRiKa13** [reply]
Really nice idea, +1

**Reply by danielroe** [reply]
Good idea 👍

**Comment by vanling** [comment]
> _https://github.com/nuxt/framework/pull/4066 Rename useMeta to useHead_
> _Rename useMeta to useHead anywhere it occurs in your code, and use app.head in your nuxt.config rather than meta. Any imports of useHead should be from #imports rather than #meta._

_For people using the lodash library and not seeing any title updates or other broken behaviour  while using the `useHead`
https://github.com/cipami/nuxt-lodash/issues/9_

**this has been [fixed](https://github.com/cipami/nuxt-lodash/pull/14) in latest version**

**Comment by baixiaoyu2997** [comment]
test with last version `bridge`, `static` change to `public`, i think this should add in breaking change

**Reply by pi0** [reply]
It should be changed back to static by default for bridge (@danielroe)

**Reply by danielroe** [reply]
@baixiaoyu2997 would you explain what you mean in a new issue on https://github.com/nuxt/bridge?

**Comment by baixiaoyu2997** [comment]
`useRuntimeConfig()` return value is `Proxy` type, and following code both can get the value
```
export default defineNuxtPlugin(() => {
  console.log(useRuntimeConfig().public.myValue)
  console.log(useRuntimeConfig().myValue)
})
```
so, what's different, Is it necessary wirte `public`

**Reply by pi0** [reply]
@baixiaoyu2997 Yes. Proxy is only for backward compatibility. Using `.public` makes code constant with new `runtimeConfig` option.

**Reply by baixiaoyu2997** [reply]
@pi0  hey, a little confused here, if for compatibility, so i should't change my code ,right? 

my original code is like this
```
publicRuntimeConfig:{
    test:1
}
```
and change in plugin : 
```js
export default defineNuxtPlugin((nuxtApp) => {
  if (!process.server) {
    const config = useRuntimeConfig();
    config.test = 2
  }
});
```
it should worked, but now ,it doesn't, if u need repo, just tell me

**Comment by adriaandotcom** [comment]
Would you mind commenting on this thread @pi0 after you add a breaking change? That way all subscribers get a notification.

**Comment by sduduzog** [comment]
@soixt Wouldn't this be a more suitable comment to ask in this repository https://github.com/nuxt-community/auth-module

**Reply by cnikolov** [reply]
This fixed it for me
    "nuxt3": "npm:nuxt3@latest",
    "nuxt": "npm:nuxt3@latest",

**Comment by lperez22** [comment]
Hey just wanted to post this here in case someone needs output files to be in mjs format 
```
nuxt.config.js

    vite: {
		build: {
		  rollupOptions: {
			output: {
			  chunkFileNames: '_nuxt/[name]-[hash].mjs',
			},
		  },
		},
    },
```
